### PR TITLE
Registered users can check out 

### PR DESCRIPTION
**Pull request by thesuss**
PT Story: 
https://www.pivotaltracker.com/story/show/131079547

Changes made by this PR:
- Create checkout page, display cart information
- Utilize session to store cart_id, clear after checkout
- Restrict checkout to registered users (the only type of user right now is Customer, will need to adjust further when we implement authorization)
- Allow for multiple items in the cart

Things I learned:
- Sessions! Using, clearing, initializing (data points on).

### CODE REVIEW

File: features/step_definitions/cart_steps.rb [hunk 1]
```diff
@@ -25,9 +25,9 @@
 end
 
 Given(/^there are no dishes in my cart$/) do
- expect(ShoppingCart.first).to eq nil
+ expect(ShoppingCart.first..shopping_cart_items[0]).to eq nil
```
Comment (tochman) [inline features/step_definitions/cart_steps.rb]: !! mind the dot.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,3 +1,14 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  before_filter :store_current_location, :unless => :devise_controller?
+
+  private
+  def store_current_location
+    store_location_for(:user, request.url)
+  end
```
Comment (tochman) [inline app/controllers/application_controller.rb:8]: Is this used to redirect user back to the previous page?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,3 +1,14 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+  before_filter :store_current_location, :unless => :devise_controller?
```
Comment (tochman) [inline app/controllers/application_controller.rb]: Am I seeing the old hash syntax? Really, still?

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
+
+  private
+  def store_current_location
+    store_location_for(:user, request.url)
+  end
+
+  def after_sign_out_path_for(resource)
+    request.referrer || root_path
+  end
```
Comment (tochman) [inline app/controllers/application_controller.rb]: Not sure if the main controller is the right place for this method. Also, why is it part of this PR, can't really see in what way it fits into the scope of this feature to handle the after sign out path for a user.

File: app/controllers/carts_controller.rb [hunk 5]
```diff
@@ -1,13 +1,31 @@
 class CartsController < ApplicationController
   def index
-    @cart = ShoppingCart.first
+    if session[:cart_id] == nil
+      @cart = nil
+    else
+      @cart = ShoppingCart.find(session[:cart_id])
+    end
```
Comment (tochman) [inline app/controllers/carts_controller.rb]: I would refactor this to a ternary operator. Not a blocker though.

File: features/customer.feature [hunk 6]
```diff
@@ -17,7 +17,7 @@ Scenario: I register as a Customer
 
 Scenario: Redirects to restaurant index page after registration
   Given I register as a user with username "Amber" and email "amber@random.com"
-  Then I should be on the "restaurant" page
+  Then I should be on the "index" page
   And I should see "Amber"
```
Comment (tochman) [inline features/customer.feature:21]: Feels out of scope of this feature.

File: features/step_definitions/user_steps.rb [hunk 7]
```diff
@@ -17,6 +17,19 @@
...
+    When I fill in:
+      | element                    | content          |
+      | Name                       | #{name}           |
+      | Email                      | #{email} |
+      | Address                    | Fjällgatan 3, Göteborg, Västa Götaland, Sverige |
+      | Password                   | password         |
+      | Password confirmation      | password         |
+    And I click the "Register" button
+  }
+end
```
Comment (tochman) [inline features/step_definitions/user_steps.rb:31]: I would prefer you guys used the Device/Warden helpers here rather than going through the interface. Not a blocker though.

File: features/customer.feature (hunk 6 above)
Comment (AmberWilkie) [reply features/customer.feature:21]: I can answer this one: we changed the path following registration for the cart feature, so we had to adjust the customer test so it would pass.

File: app/controllers/carts_controller.rb [hunk 8]
```diff
@@ -1,22 +1,11 @@
 class CartsController < ApplicationController
   def index
-    if session[:cart_id] == nil
-      @cart = nil
-    else
-      @cart = ShoppingCart.find(session[:cart_id])
-    end
+    find_or_create_cart
```
Comment (tochman) [inline app/controllers/carts_controller.rb]: Good job. You can probably add this as an `before_action` if called from two methods (index and show)

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (thesuss) [reply app/controllers/application_controller.rb]: True, this was not necessary. I followed a guide a little too far.

Comment (tochman) [reply app/controllers/application_controller.rb]: Okay, thx for the clarification

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (thesuss) [reply app/controllers/application_controller.rb:8]: Yes, I know this is the old way, but I couldn't make it work the new way. 
So I added a chore in Piv to change it when time allows.

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (thesuss) [reply app/controllers/application_controller.rb]: Sorry, fixed.

**Comment by AmberWilkie** [comment]
Changes made by this PR:
- Create checkout page, display cart information
- Utilize session to store cart_id, clear after checkout
- Restrict checkout to registered users (the only type of user right now is Customer, will need to adjust further when we implement authorization)
- Allow for multiple items in the cart

Things I learned:
- Sessions! Using, clearing, initializing (data points on).

**Comment by tochman** [comment]
Please update the PR description. I assume that since the WIP tag is removed I can do a final review and merge?

**Comment by AmberWilkie** [comment]
We feel it's "done".

**Comment by tochman** [comment]
I would like you to examine this code and compare it to your solution:

``` ruby
class CartsController < ApplicationController
  before_action :find_or_create_cart, only: [:index, :add_item]

  def index
  end

  def add_item
    @dish = Dish.find(params[:dish_id])
    @cart.add(@dish, @dish.dish_price)
    redirect_back(fallback_location: restaurant_path)
  end

  def checkout
    @order = ShoppingCart.find(params[:format])
    session.delete(:cart_id)
    flash[:success] = 'Your food is on its way!'
  end

  private
  def find_or_create_cart
    @cart = ShoppingCart.find_or_create_by(id: session[:cart_id])
    session[:cart_id] = @cart.id
  end
end
```

**Comment by AmberWilkie** [comment]
Ok that makes sense. I didn't get from the research I did on find_or_create_by that we could make something from nil (because `session[:cart_id]` is `nil` in some of our tests). Nothing breaks when adding in this code.

File: app/controllers/carts_controller.rb [hunk 9]
```diff
@@ -14,17 +16,13 @@ def checkout
     @order = ShoppingCart.find(params[:format])
     session.delete(:cart_id)
     flash[:success] = "Your food is on its way!"
```
Comment (tochman) [inline app/controllers/carts_controller.rb:16]: And I really want you to only use double quotes when needed.
